### DISCUSSION TITLE
How can I treat minors as majors?

### DISCUSSION THREAD

**Discussion by TWiStErRob** [renovatebot/renovate#14919] [Request Help]
I like the behavior that minors are grouped together, but for some libraries minor *is* major.

For example Kotlin has these versions:
1.4.10 => 4.1.0
1.4.30 => 4.3.0
1.4.31 => 4.3.1
1.4.32 => 4.3.2
1.5.0  => 5.0.0
1.5.10 => 5.1.0
1.5.20 => 5.2.0
1.5.21 => 5.2.1
1.5.30 => 5.3.0
1.5.31 => 5.3.1
1.5.32 => 5.3.2

but in reality practically, considering the compatibilities and the deprecations, they are actually as stated after the arrow above.

How can I achieve the above mapping regarding major/minor/patch differences for specific libraries?

example behavior:
If I'm on 1.4.10,
I'd like to see a change for 1.4.32 included in "Update all non-major dependencies",
and I'd like to see a PR for updating to 1.5.32 as a separate PR.

---

Note: this is really similar how Java started out as 1.1, 1.2, ..., 1.5=5, 1.6=6, 1.7=7, 1.8=8, 9, 10, 11, ...

**Comment by rarkins** [comment]
Minor is never major, so the short answer to your question is "No". You sound like you have a valid requirement, but you're framing it in terms of asking for a *solution* which is invalid (minor != major). If want to ask about how to exclude certain packages from groups, reframe it that way. And perhaps separateMinorPatch is relevant to you too.

**Reply by TWiStErRob** [reply]
> You sound like you have a valid requirement

I think you understood me, the requirement is real because the Kotlin and Android Gradle Plugin versions always contain breaking changes, any update to the second number in the version breaks the build on most real projects and needs adjustments. So from my end-user perspective they're major updates even if the semver numbering interpretation tells otherwise.

> If want to ask about how to exclude certain packages from groups, reframe it that way.

I don't have enough experience with renovatebot to understand this terminology yet. I asked "treat minor as major" because there considerations for majors, for example automerging and separation (and I'm sure there's more). It looks like `minor: { automerge: true }, major: { automerge: true }` is possible, but `separateMultipleMajor` is implemented in a different way so I am not able to say `separateMultipleMinor`. So that might be an additional feature request, but it's unnecessary if I can just say, that a package is not following "major.minor.patch" semver, but instead following "epic.major.minor" versioning. I understand this might be an xyproblem, and ask you to please help reframe it so I can reach my goals.

I played around and got this far:
```json
{
	"extends": [
		"config:base"
	],
	"baseBranches": ["renovate/configure"],
	"commitMessageExtra": "from {{{currentValue}}} to {{{newValue}}} ({{#if isMajor}}major{{else}}notmajor{{/if}}/{{#if isSingleVersion}}single{{else}}notsingle{{/if}})",
	"packageRules": [
		{
			"groupName": "Kotlin",
			"matchPackageNames": [
				"org.jetbrains.kotlin:kotlin-stdlib",
				"org.jetbrains.kotlin.android"
			],
			"separateMajorMinor": true,
			"separateMultipleMajor": true,
			"separateMinorPatch": true
		}
	]
```
where the first update is good, that's progress based on your `separateMinorPatch` suggestion, but the second one goes too far. I don't want to jump over "major" updates, 1.4 -> 1.5 -> 1.6.
![image](https://user-images.githubusercontent.com/2906988/162574900-1d40e49f-6316-4409-925c-039fafab129b.png)

I have a feeling https://docs.renovatebot.com/configuration-options/#minor might be somehow a potential solution, but I don't see how.

**Comment by thejan2009** [comment]
You can redefine versioning for packages, but take into account there will be issues when there actually is a major update.
```json
{
  "packageRules": [{
    "matchPackageNames": ["org.jetbrains.kotlin*"],
    "versioning": "regex:^v\\d\\.?(?<major>\\d+)\\.(?<minor>\\d+)$"
  }]
}
```

**Reply by TWiStErRob** [reply]
_(Note: for debugging I'm using `"commitMessageExtra": "from {{{currentValue}}} to {{{newValue}}} ({{#if isMajor}}major{{else}}notmajor{{/if}}/{{#if isSingleVersion}}single{{else}}notsingle{{/if}})"`)_

Interesting one, I gave it a try: `regex:^(?<major>\\d+\\.\\d+)\\.(?<minor>\\d)-?(?<patch>\\d)?$`, but that failed miserably:
```
Update Kotlin from 1.4.32 to 1.3.72 (notmajor/single) (minor)
```
It looks like `1.4` and `1.3` is getting converted to a number before comparison so `"1.4" == 1` and `"1.3" == 1` and hence the minor "7" wins (that was the highest one, ever). I tried this in hopes that it's a string-comparison and in that case "actually a major update" wouldn't be a problem.

---

Using a hardcoded `1\\.(?<major\\d+)` works like a charm though:
```
Update Kotlin from 1.4.32 to 1.5.32 (major/single) (major)
Update Kotlin from 1.4.32 to 1.6.20 (major/single) (major)
```

**Reply by TWiStErRob** [reply]
So I'm happy with this, but sadly this only works for Kotlin, which has been in the 1.x version for 7 years now.

The other library, Android Gradle Plugin, which has a similar thing going on, would really need the support of string comparison to work:
```
(?<major>\\d+\\.\\d+)\\.(?<minor>\\d)
```
because in that case "actual major updates" have happened multiple times already:
..., 3.4.x, 3.5.x, 3.6.x, 4.0.x, 4.1.x, 4.2.x, 7.0.x, 7.1.x, 7.2.x

**Reply by viceice** [reply]
i think we can probably allow / parse capture groups as float, so regex versioning can match those case as expected. 🤔 should also be non breaking.

@rarkins WDYT?

**Reply by TWiStErRob** [reply]
😱 1.4 < 1.12 should be true when it comes to versioning. Float would be good enough for most cases, but at the same time inherently broken. I was aware of that when trying this, and was hoping that Kotlin 1.9 in a year or two might be followed by 2.0, but that's not necessarily true. Float could still be an improvement, but there might be another solution that's more dynamic and correct.

**Reply by TWiStErRob** [reply]
Would it be too much to do a threeway comparison?
```
when(major) {
     matches \d+ -> use existing logic
     matches \d+(\.\d+)+ -> split and do lexicographic comparison
     else -> use string comparison
}
```
(First two might be combined with `*` because a lexicographic comparison on a single item is equivalent to current behaviour.)

**Comment by rarkins** [comment]
Related: #6576 (separateMultipleMinor) and #1257 (touches on the concept of a "isBreaking" classification which can be declared independently of major version updates)

**Comment by HonkingGoose** [comment]
@TWiStErRob Maybe you can use the "Dependency Dashboard Approval" workflow for the Kotlin package, instead of creating complicated rules? I've linked the relevant docs below:

- [Use cases, Dependency Dashboard Approval](https://docs.renovatebot.com/getting-started/use-cases/#dependency-dashboard-approval)

- [Renovate FAQ, Tell Renovate to ask for approval before creating a Pull Request](https://docs.renovatebot.com/faq/#tell-renovate-to-ask-for-approval-before-creating-a-pull-request)

**Reply by TWiStErRob** [reply]
My reason for enabling renovate in the first place is that I get email notifications of PRs that are already built by CI (assuming I read personal emails after work ;). With this workflow I don't see how that works yet.

**Reply by TWiStErRob** [reply]
Another thing is that I want to be able to auto-create and auto-merge safe "minor" and "patch" updates (when the last digit of 1.5.31 to 1.5.32 is changed).

**Reply by HonkingGoose** [reply]
## Renovate config

How about something like this then?

```json
{
  "packageRules": [
    {
      "description": ["Require dashboard approval for major Kotlin updates"],
      "matchPackageNames": ["kotlin"],
      "matchUpdateTypes": ["major"],
      "dependencyDashboardApproval": true,
      "prCreation": "not-pending"
    },
    {
      "description": [
        "Raise PRs for minor and patch Kotlin updates and automerge them if tests pass"
      ],
      "matchPackageNames": ["kotlin"],
      "matchUpdateTypes": ["minor", "patch"],
      "automerge": true
    }
  ]
}
```

## Steps for `major` Kotlin updates

1. You get a Dependency Dashboard issue
2. You click on the "Kotlin" update checkbox for Major update
3. Renovate creates the Kotlin branch, and lets the tests run
4. After _all_ tests are done - because of `"prCreation": "not-pending"` - Renovate creates the PR and you get a mail from GitHub

## Steps for minor and patch Kotlin updates:

1. Do nothing, wait for Renovate to create branches and run the tests
2. Get PRs for updates
3. The PR will get automerged after a while if all tests pass

**Reply by TWiStErRob** [reply]
Re email: I need to take action (2) to get email (4), but maybe (1) also sends email?

Re major/minor split: I still don't see how this helps with the original issue, which is that what you call "minor" are major updates, what you call "patch" are actually minor AND patch updates, see mapping in OP. Although I think I see what you're getting at, since my "major" interpretation actually matches Renovate's minor; so if I move minor to dashboard that could help.

**Reply by HonkingGoose** [reply]
> Re email: I need to take action (2) to get email (4), but maybe (1) also sends email?

(1) The dependency dashboard will result in just _one_ email if you're "watching" the repository on GitHub. You'll get that email when you enable the dashboard, because it creates a issue for the dashboard. Afterwards you'll get no more emails about the dashboard. 😉 

> Re major/minor split: I still don't see how this helps with the original issue, which is that what you call "minor" are major updates, what you call "patch" are actually minor AND patch updates, see mapping in OP. Although I think I see what you're getting at, since my "major" interpretation actually matches Renovate's minor; so if I move minor to dashboard that could help.

My idea was to take `major` updates manually, as both you and the Kotlin maintainers agree that it will break stuff. For all other updates, let the bot try it out for you and see what happens.

It's all really complicated stuff that you're trying to do. I was trying to get back to a simpler and "conventional" config, that doesn't try to change the Renovate bot behavior.

Have you raised any issue with the maintainers of Kotlin and Android Gradle Plugin to discuss their versioning process? If the maintainers tagged breaking releases as a new `major` instead of putting it in `minor` or `patch`, then the whole thing would be fixed. But I'm probably over-simplifying complex decisions by the maintainers here. 😉

**Reply by TWiStErRob** [reply]
There was no "major" update of Kotlin in 7 years, so yes with the `versioning` regex I could take majors manually and do minors by automation. The problem is that patches can definitely just work, and minors maybe, but they cannot be separate because they're represented by the same number.

Re changing, not sure if they can. AGP just changed versioning to follow Gradle's majors. Kotlin versioning I'm not sure, [asked](https://kotlinlang.slack.com/archives/C0B9K7EP2/p1650647460034829), let's see where it leads. Edit: lead to https://youtrack.jetbrains.com/issue/KT-44136
